File: FrontEnd/insuranceApp/src/app/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product } from './product';

@Injectable({
  providedIn: 'root',
})
export class ProductService {
  private baseUrl = 'http://localhost:9093/products';

  constructor(private http: HttpClient) {}

  createProduct(product: Product): Observable<Product> {
    return this.http.post<Product>(`${this.baseUrl}/add`, product);
  }

  getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.baseUrl}/view/all`);
  }

  updateProduct(productId: string, product: Product): Observable<Product> {
    return this.http.put<Product>(
      `${this.baseUrl}/update/${productId}`,
      product
    );
  }

  deleteProduct(productId: string): Observable<string> {
    return this.http.delete<string>(`${this.baseUrl}/delete/${productId}`);
  }

  getProductByProductId(productId: string): Observable<Product> {
    return this.http.get<Product>(`${this.baseUrl}/view/${productId}`);
  }

  getProductsByCompanyId(companyId: string): Observable<Product[]> {
    return this.http.get<Product[]>(
      `${this.baseUrl}/view/products-by-company/${companyId}`
    );
  }

  getProductsByCategory(category: string): Observable<Product[]> {
    return this.http.get<Product[]>(
      `${this.baseUrl}/view/products-by-category/${category}`
    );
  }

  deleteProductsByCompanyId(companyId: string): Observable<string> {
    return this.http.delete<string>(
      `${this.baseUrl}/delete/products-by-company/${companyId}`
    );
  }

  viewAvailableProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.baseUrl}/view/products-available`);
  }

  viewAvailableProductsByCompanyId(companyId: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.baseUrl}/view/all/${companyId}`);
  }

  setProductAvailableByProductId(productId: string): Observable<string> {
    return this.http.put<string>(
      `${this.baseUrl}/update/setProductAvailability/${productId}`,
      null
    );
  }

  getProductsByLocationName(locationName: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.baseUrl}/location/${locationName}`);
  }
}

File: BackEnd/productMicroservice/src/main/java/com/cgi/productMicroservice/service/ProductService.java
package com.cgi.productMicroservice.service;

import java.util.List;

import com.cgi.productMicroservice.exception.LocationNotFoundException;
import com.cgi.productMicroservice.exception.ProductAlreadyExistsException;
import com.cgi.productMicroservice.exception.ProductNotFoundException;
import com.cgi.productMicroservice.model.Product;

public interface ProductService {
	
	
	//create Product
	public Product createProduct(Product product) throws ProductAlreadyExistsException;
	
	//update Product
	public Product updateProduct(String productID, Product product) throws ProductNotFoundException;
	
	//delete Product
	public boolean deleteProduct(String productID) throws ProductNotFoundException;
	
	//get a Product by productID
	public Product getProductByProductID(String productID) throws ProductNotFoundException;
	
	//get all Products
	public List<Product> getAllProducts();
	
	//get all products by companyID
	public List<Product> getAllProductsByCompanyID(String companyId);
	
	//get all products by category
	public List<Product> getAllProductsByCategory(String category);
	
	//delete all products by companyID
	public void deleteByCompanyId(String companyId);
	
	//get all products by availability(isAvailable=true)
	public List<Product> viewAvailableProducts();
	
	//get all products by availability(isAvailable=true) and by companyId
	public List<Product> viewAvailableProductsByCompanyId(String companyId);
	
	//make a product available by productId
	public Product setProductAvailableByProductId(String productId) throws ProductNotFoundException;
	
	//get all products by locations
	public List<Product> getProductsByLocationName(String locationName) throws LocationNotFoundException;
}
